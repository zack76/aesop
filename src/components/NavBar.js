import React, { Component } from 'react';

const NavBar = () => (
    <nav className="navbar navbar-light bg-light text-center">
        <a className="navbar-brand" href="#">Aesop</a>
    </nav>
)

export default NavBar
