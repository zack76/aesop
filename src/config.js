export const Config = {
    CROS_ANYWHERE : 'https://cors-anywhere.herokuapp.com/',
    AESOP_URL : 'https://www.aesop.com/au/api/v1/nav/shop/',

};