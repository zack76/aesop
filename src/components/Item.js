import React  from 'react';
import 'bootstrap/dist/css/bootstrap.min.css';
import { Row } from "react-bootstrap";
import { Config } from '../config';
import { connect } from 'react-redux';
import { selectProduct } from '../reducers/selectedProduct/selectProduct';


const Item = ({ items, selectProduct }) => {
    if (items && items.length > 0) {
        let itemsUnderRows = []
        items.forEach((category, index) => {
            if (index % 4 === 0) {
                itemsUnderRows.push([])
            }
            const lastRow =
                itemsUnderRows[itemsUnderRows.length - 1]
            lastRow.push(category)
        })
        return itemsUnderRows.map((itemsUnderRow, rowIndex) => (
            <Row key={rowIndex} className="row-with-margin">
                {itemsUnderRow.map((tmpProduct, itemIndex) => {
                    let product = tmpProduct
                    return (
                        <div
                            key={itemIndex}
                            className="row-item col-xs-12 col-sm-4 col-md-3 card-item"
                            onClick={() => selectProduct(product)}
                        >
                            <div key={product.name} className="image-container">
                                <img className="image-contained"
                                     id={'image-' + product.name}
                                     src={Config.AESOP_URL + (product.thumbnail || product.imageUrl)}
                                     alt={product.name}
                                />
                                <div className="item-desc">
                                    <strong>{product.name}</strong>
                                    <br/>
                                    <div className="item-description">{product.shortDescription}</div>
                                </div>
                            </div>
                        </div>
                    )
                })}
            </Row>
        ))
    } else return null;
}

export const mapStateToProps = ({ selectedProduct }) => {
    return {
        selectedProduct: selectedProduct
    }
}

export default connect(mapStateToProps , { selectProduct })(Item);
