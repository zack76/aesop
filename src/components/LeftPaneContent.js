import React, { Component } from 'react';
import {Accordion, Button, Card, Table, Row} from "react-bootstrap";
import Item from "./Item";

const renderVariants = (variants) => {
    return variants.map(variant => {
        return (
            <Accordion key={variant.name || variant.sku} defaultActiveKey="0">
                <Card>
                    <Card.Header>
                        <Accordion.Toggle as={Button} variant="link"
                                          eventKey={'variant' + variant.name || variant.sku}>
                            {variant.name || variant.sku}
                        </Accordion.Toggle>
                    </Card.Header>
                    <Accordion.Collapse eventKey={'variant' + variant.name || variant.sku}>
                        <Card.Body>
                            {renderVariant(variant)}
                        </Card.Body>
                    </Accordion.Collapse>
                </Card>
            </Accordion>
        )
    })
}

const renderVariant = (variant) => {
    var keys = Object.keys(variant)
    var values = Object.values(variant)
    return (
        <Table striped bordered hover size="sm">
            <thead>
            <tr>
                <th>Variant</th>
                <th>Value</th>
            </tr>
            </thead>
            <tbody>
            {keys.map((key, index) => (
                <tr key={key}>
                    <td>{key}</td>
                    <td>{typeof values[index] === "boolean" ? values[index].toString() : values[index]}</td>
                </tr>
            ))}
            </tbody>
        </Table>
    )
}

const LeftPaneContent = ({product}) => {
    if (product) {
        return  (
            <div>
                <h1>{product.name}</h1>
                {renderVariants(product.variants)}
            </div>
        )
    } else return null;
}

export default LeftPaneContent
