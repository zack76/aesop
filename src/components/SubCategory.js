import React from 'react';
import 'bootstrap/dist/css/bootstrap.min.css';
import {Accordion, Card, Button, Col, Row} from "react-bootstrap";
import Item from "./Item"


const SubCategory = ({ subcategories }) => {
    return subcategories.map((subcategory, index) =>
        (
            <Accordion key={subcategory.name + index} defaultActiveKey="0">
                <Card>
                    <Card.Header>
                        <Accordion.Toggle as={Button} variant="link"
                                          eventKey={'subcategory' + subcategory.name + index}>
                            {subcategory.name}
                        </Accordion.Toggle>
                    </Card.Header>
                    <Accordion.Collapse eventKey={'subcategory' + subcategory.name + index}>
                        <Card.Body>
                            {subcategory.items[0].price ?
                                <Col xs={12} key={subcategory.name + index}>
                                    <Row>
                                        <Item items={subcategory.items}/>
                                    </Row>
                                </Col>
                                :
                                <SubCategory key={'subcategory' + subcategory.name + index}
                                subcategories={subcategory.items}/>
                            }
                        </Card.Body>
                    </Accordion.Collapse>
                </Card>
            </Accordion>
        ))
}

export default SubCategory
