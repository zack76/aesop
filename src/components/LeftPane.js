import React  from 'react';
import { connect } from 'react-redux';
import SlidingPanel from 'react-sliding-side-panel';
import 'react-sliding-pane/dist/react-sliding-pane.css';
import { removeSelectProduct } from "../reducers/selectedProduct/removeSelectProduct";
import LeftPaneContent from "../components/LeftPaneContent";
import FontAwesome from 'react-fontawesome';

const LeftPane = ({selectedProduct, removeSelectProduct}) => (
    <SlidingPanel
        type={'left'}
        isOpen={selectedProduct !== null && selectedProduct !== undefined}
        size={30}
        className="left-pane"
    >
        <div>
            <div className="btn" onClick={() => {removeSelectProduct()}}>
                <FontAwesome name="times" size="2x"/>
            </div>
            <LeftPaneContent product={selectedProduct} />
        </div>
    </SlidingPanel>
)


const mapStateToProps = ({ products, selectedProduct }) => {
    return {
        products: products,
        selectedProduct: selectedProduct
    }
}

export default connect(mapStateToProps , { removeSelectProduct })(LeftPane);
