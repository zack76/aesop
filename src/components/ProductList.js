import React, { Component } from 'react';
import 'bootstrap/dist/css/bootstrap.min.css';
import {Accordion, Card, Button} from "react-bootstrap";
import SubCategory from "./SubCategory"

const ProductList = ({ products }) => {
    return products.map((category, index) => (
        <Accordion key={index} defaultActiveKey="0">
            <Card>
                <Card.Header>
                    <Accordion.Toggle as={Button} variant="link" eventKey={'category' + index}>
                        {category.name}
                    </Accordion.Toggle>
                </Card.Header>
                <Accordion.Collapse eventKey={'category' + index}>
                    <Card.Body>
                        <SubCategory subcategories={category.items} />
                    </Card.Body>
                </Accordion.Collapse>
            </Card>
        </Accordion>
    ))
}

export default ProductList
