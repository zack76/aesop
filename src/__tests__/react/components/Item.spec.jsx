import React from 'react';
import { mount, shallow } from 'enzyme';
import { Item, mapStateToProps } from '../../../components/Item';
import { Row } from "react-bootstrap";
import { Config } from '../../../config';


describe('Item', () => {
    const props = {
        selectProduct: jest.fn(),
        items: [
            { name: 'toy', thumbnail: 'fake_src_1', imageUrl: 'imageUrl', shortDescription: 'desc', variants: [{'name': 'a', 'B': 'b'}] },
            { name: 'paper', thumbnail: null, imageUrl: 'imageUrl', shortDescription: 'desc', variants: [{'name': 'a', 'B': 'b'}] },
            { name: 'food', thumbnail: 'fake_src_3', imageUrl: 'imageUrl', shortDescription: 'desc', variants: [{'name': 'a', 'B': 'b'}] }
        ]
    };

    const porpsWithNoItems = {
        items: []
    }

    const initialState = {
        selectedProduct: 1
    };

    it('Should render items without error', () => {
        const wrapper = mount(<Item {...props} />);
        expect(wrapper.find(Row).length).toEqual(Math.ceil(props.items.length / 4));
        expect(wrapper.find('.card-item').length).toBe(3);
        expect(wrapper.find('img').length).toBe(3);
        expect(wrapper.find('img').first().prop('src')).toBe(Config.AESOP_URL + 'fake_src_1');
        expect(wrapper.find('img').first().prop('id')).toBe('image-toy');
        expect(mapStateToProps(initialState).selectedProduct).toEqual(1);
    })

    it('Should render items with no thumbnail without error', () => {
        const props = {
            selectProduct: jest.fn(),
            items: [
                { name: 'toy', thumbnail: null, imageUrl: 'imageUrl', shortDescription: 'desc', variants: [{'name': 'a', 'B': 'b'}] },
            ]
        };
        const wrapper = mount(<Item {...props} />);
        expect(wrapper.find('img').first().prop('src')).toBe(Config.AESOP_URL + 'imageUrl');
        expect(mapStateToProps(initialState).selectedProduct).toEqual(1);
    })

    it('Should render no item without error', () => {
        const wrapper = mount(<Item {...porpsWithNoItems} />);
        expect(wrapper.children().length).toBe(0);
    })

    it('should render without error', () => {
        const wrapper = shallow(<Item debug />);
        expect(wrapper).toMatchSnapshot();
    });

    it('Should item be clicked', () => {
        const wrapper = mount(<Item {...props} />);
        expect(props.selectProduct.mock.calls.length).toBe(0);
        wrapper
            .find('.card-item')
            .first()
            .simulate('click');

        expect(props.selectProduct.mock.calls.length).toBe(1);
    });

});
