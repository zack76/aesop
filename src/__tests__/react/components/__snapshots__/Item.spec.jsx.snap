// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item should render without error 1`] = `""`;
